File: crates/generate/ess-synth/tests/fixtures/selection-native.go
package system_test

import (
	"example.invalid/selection/system"
	"example.invalid/selection/types/core"
	"testing"
)

func leg(id string, domain core.Domain, source core.Source) *core.Leg {
	value := &core.Leg{Id: id, From: "remote"}
	if domain != nil {
		value.Domain = &domain
	}
	if source != nil {
		value.Source = &source
	}
	return value
}
func text(value string) *string { return &value }
func same(a, b *string) bool {
	if a == nil || b == nil {
		return a == b
	}
	return *a == *b
}
func TestMixedOccurrencesAndFallbacks(t *testing.T) {
	I, E, W := core.DomainInternal{}, core.DomainExternal{}, core.SourceWebrtc{}
	rows := []struct {
		data            []*core.Leg
		agent, external *string
	}{
		{[]*core.Leg{}, nil, nil},
		{[]*core.Leg{nil, leg("x", nil, nil)}, text("x"), text("x")},
		{[]*core.Leg{leg("x", E, nil), leg("y", I, nil)}, text("y"), text("x")},
		{[]*core.Leg{leg("x", E, W)}, text("x"), text("x")},
		{[]*core.Leg{leg("x", E, W), leg("y", E, nil)}, text("x"), text("y")},
		{[]*core.Leg{leg("x", I, nil), leg("y", E, W)}, text("x"), text("y")},
		{[]*core.Leg{leg("x", E, W), leg("y", E, W)}, text("x"), text("y")},
		{[]*core.Leg{leg("x", E, nil), leg("y", E, nil)}, text("x"), text("x")},
		{[]*core.Leg{nil, leg("", nil, nil), leg("z", nil, nil)}, text("z"), text("z")},
		{[]*core.Leg{leg("", I, nil), leg("y", E, nil)}, text(""), text("y")},
		{[]*core.Leg{leg("", E, nil), leg("y", nil, nil)}, text("y"), text("")},
		{[]*core.Leg{leg("x", I, nil), leg("y", I, nil), leg("z", E, nil)}, text("x"), text("z")},
		{[]*core.Leg{leg("same", E, W), leg("same", E, nil)}, text("same"), text("same")},
	}
	for index, row := range rows {
		value, failure := system.Choose(core.Arrived{Data: row.data})
		if failure != nil {
			t.Fatalf("row%d: %v", index, failure)
		}
		if !same(value.AgentId, row.agent) || !same(value.ExternalId, row.external) {
			t.Fatalf("row%d wrong occurrences", index)
		}
	}
}
func TestOverLimitTail(t *testing.T) {
	data := make([]*core.Leg, 65)
	for index := range data {
		data[index] = leg("x", core.DomainInternal{}, nil)
	}
	_, failure := system.Choose(core.Arrived{Data: data})
	if failure == nil || failure.Cause != system.SelectionResource {
		t.Fatalf("expected resource refusal: %v", failure)
	}
}
func TestMalformedEnumTail(t *testing.T) {
	var invalid core.Domain
	tail := leg("bad", nil, nil)
	tail.Domain = &invalid
	_, failure := system.Choose(core.Arrived{Data: []*core.Leg{leg("x", core.DomainInternal{}, nil), tail}})
	if failure == nil || failure.Cause != system.SelectionInvalidInput || failure.Index == nil || *failure.Index != 1 {
		t.Fatalf("expected typed malformed tail refusal: %v", failure)
	}
}

File: crates/generate/ess-synth/tests/fixtures/selection-native.rs
use selection_system::{choose, SelectionFailureCause};
use selection_types::core::{Arrived, Domain, Leg, Source};
fn leg(id: &str, domain: Option<Domain>, source: Option<Source>) -> Option<Leg> { Some(Leg { id:id.into(),domain,source,from:"remote".into() }) }
#[test]
fn mixed_occurrences_and_fallbacks() {
    use Domain::{Internal as I, External as E}; use Source::Webrtc as W;
    let rows = [
        (vec![],None,None),
        (vec![None,leg("x",None,None)],Some("x"),Some("x")),
        (vec![leg("x",Some(E),None),leg("y",Some(I),None)],Some("y"),Some("x")),
        (vec![leg("x",Some(E),Some(W))],Some("x"),Some("x")),
        (vec![leg("x",Some(E),Some(W)),leg("y",Some(E),None)],Some("x"),Some("y")),
        (vec![leg("x",Some(I),None),leg("y",Some(E),Some(W))],Some("x"),Some("y")),
        (vec![leg("x",Some(E),Some(W)),leg("y",Some(E),Some(W))],Some("x"),Some("y")),
        (vec![leg("x",Some(E),None),leg("y",Some(E),None)],Some("x"),Some("x")),
        (vec![None,leg("",None,None),leg("z",None,None)],Some("z"),Some("z")),
        (vec![leg("",Some(I),None),leg("y",Some(E),None)],Some(""),Some("y")),
        (vec![leg("",Some(E),None),leg("y",None,None)],Some("y"),Some("")),
        (vec![leg("x",Some(I),None),leg("y",Some(I),None),leg("z",Some(E),None)],Some("x"),Some("z")),
        (vec![leg("same",Some(E),Some(W)),leg("same",Some(E),None)],Some("same"),Some("same")),
    ];
    for (index,(data,agent,external)) in rows.into_iter().enumerate() {let result=choose(&Arrived{data}).unwrap();assert_eq!(result.agent_id.as_deref(),agent,"agent row {index}");assert_eq!(result.external_id.as_deref(),external,"external row {index}");}
}
#[test]
fn oversized_tail_refuses_before_returning_input() {let data=vec![leg("x",Some(Domain::Internal),None);65];let failure=choose(&Arrived{data}).unwrap_err();assert_eq!(failure.cause,SelectionFailureCause::Resource);assert_eq!(failure.input,0);}
#[test]
fn examined_oversized_text_tail_is_not_hidden_by_early_match() {let data=vec![leg("x",Some(Domain::Internal),None),leg(&"x".repeat(4097),None,None)];let failure=choose(&Arrived{data}).unwrap_err();assert_eq!(failure.cause,SelectionFailureCause::Resource);assert_eq!(failure.index,Some(1));}
